Taar similarity job

Dexterp37 opened the pull request:
This job is used to compute TAAR similarity-based addon donor list.

This job clusters users into different groups based on their active addons. A representative users sample is selected from each cluster ("donors") and is saved to a model file along with a feature vector that will be used, by the TAAR library module, to perform recommendations.

*Please do not review yet*

Dexterp37 commented:
@mlopatka, can you take an high level look to check that the similarity based module behaves as expected? Please also check that the test is testing the right thing.

mlopatka commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,163 @@
+        .where("active_addons IS NOT null")\
+        .where("size(active_addons[0]) > 2")\
+        .where("size(active_addons[0]) < 100")\
+        .where("normalized_channel = 'release'")\
```
Don't we also want:
` AND build IS NOT NULL AND build[0].application_name='Firefox'`

Dexterp37 replied:
Good catch!

mlopatka commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,163 @@
+    # steps work.
+    hashing_stage = HashingTF(inputCol="addon_ids", outputCol="hashed_features")
+    idf_stage = IDF(inputCol="hashed_features", outputCol="features", minDocFreq=1)
+    bkmeans_stage = BisectingKMeans(k=num_clusters)
```
Not necessary now, but a future imporvement may be to have the termination condition relate to `minDivisibleClusterSize`. Since Bisecting K-means is a divisive clustering method, we may want to continue to sub-divide the data until the smallest single cluster contains a minimum number of elements.

Dexterp37 replied:
Yup, I've added a comment line there.

mlopatka commented on tests/test_taar_similarity.py:
```diff
@@ -0,0 +1,213 @@
+    # 3 clusters out of this and 10 donors.
+    donors = taar_similarity.get_donors(spark, 3, 10)
+
+    # assert len(donors) == 10
```
There should be an assertion that the number of donors is within a particular margin of the requested number. It will not be equal due to the sampling from variable sized donor candidates (accross clusters) however the number of donors should always be within 2*(n) where n is the number of terminal clusters.

Dexterp37 replied:
Good point. However, to make tests less flaky, I've slightly changed the code to pass around a fixed "random seed" to the Spark ML functions to make tests reproducible. Since we generate the fake data, this allows us to precisely assert how many donors we see.

mlopatka replied:
Ah yeah, good point. For generated data it is a deterministic process, given a fixed random seed. Maybe we should add logging in the live module that warns us if this condition is not met so we can assess variability in with the cluster sizes as new client data is pulled in.

mlopatka commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,181 @@
+                                    **sampling_kwargs)
+    # Get the specific number of donors for each cluster and drop the
+    # predicted cluster number information.
+    current_sample_size = donor_df.count()
```
Can we add logging here to check if the actual number of returned donors is close to the requested number of donors?

Dexterp37 replied:
@mlopatka yes, I will add the logging line here. I have a question though: should we make the job fail if we don't get enough donors?

mlopatka replied:
As we just discusssed, we need a hard-coded failure condition for the extreme case.
The threshold should be set to <100.

Likewise, if a number of donoros less than 100 is requested at any funciton call (as input param) the job shoudl fail as well.

mlopatka commented:
@acmiyaguchi Can you take a look at this PR for the final r+?

Dexterp37 commented:
@mlopatka @acmiyaguchi this is not yet ready for review.

Dexterp37 commented:
Please note that I could now write test coverage for the lr -curves function: it consistently timed out locally in the test unless I increased the driver memory to over 2 GB.

mlopatka commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,321 @@
+     * Load same similarity function that will be used in TAAR module.
+     * Iterate through each cluster and compute in-cluster similarity.
+     * Iterate through each cluster and compute out-cluster similarity.
+     * Compute KDE density per similarity score.
```
nit: KDE can be written 

> Kernel Density Estimate

then we can drop the redundant "density"

mlopatka commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,321 @@
+    j_c = (distance.canberra(x_continuous_features, y_continuous_features))
+
+    # Take the product of similarities to attain a univariate similarity score.
+    # Add a minimal constant to prevent zero values.
```
nit: to prevent zero values `from categorical features`.

acmiyaguchi commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,321 @@
+    for each client along with the list of active add-on GUIDs.
+    """
+    def is_valid_addon(addon):
+        if addon.is_system or\
```
Nit: You could wrap this in parens and return the negation.

acmiyaguchi commented on mozetl/taar/taar_similarity.py:
```diff
@@ -0,0 +1,321 @@
+        .map(lambda p: (p["client_id"],
+             [guid for guid, data in p["active_addons"].items() if is_valid_addon(data)]))
+        .filter(lambda p: len(p[1]) > minimum_addons_count)
+        .toDF()
```
I think you can pass the column names into `.toDF`

Dexterp37 commented:
@acmiyaguchi thanks for the review. I've addressed your points and also fixed the `similarity_function` to be more robust against missing values in the data. Moreover, since the locale PR landed, I moved the whitelist code out of that module and used it to filter the addons in this module as well. Can you take another look at the changed parts?

acmiyaguchi commented:
@Dexterp37 The changes look good, looks like all of your changes have all the interesting cases being exercised by relevant tests.
